SwiftReflector/GenericReferenceRenamer.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using Dynamo;
+using SwiftReflector.SwiftXmlReflection;
+
+namespace SwiftReflector {
+	public class GenericReferenceRenamer {
+		BaseDeclaration context;
+		Func<int, int, string> genericNamer;
+		public GenericReferenceRenamer (BaseDeclaration context, Func<int, int, string> genericNamer)
+		{
+			this.context = Exceptions.ThrowOnNull (context, nameof (context));
+			this.genericNamer = Exceptions.ThrowOnNull (genericNamer, nameof (genericNamer));
+		}
+
+		public TypeSpec Rename (TypeSpec type)
+		{
+			bool changed;
+			var newType = Rename (type, out changed);
+			return changed ? newType : type;
+		}
+
+		TypeSpec Rename (TypeSpec type, out bool changed)
+		{
+			if (TypeSpec.IsNullOrEmptyTuple (type)) {
+				changed = false;
+				return type;
+			}
+			switch (type.Kind) {
+			case TypeSpecKind.Closure:
+				return Rename (type as ClosureTypeSpec, out changed);
+			case TypeSpecKind.Named:
+				return Rename (type as NamedTypeSpec, out changed);
+			case TypeSpecKind.Tuple:
+				return Rename (type as TupleTypeSpec, out changed);
+			case TypeSpecKind.ProtocolList:
+				return Rename (type as ProtocolListTypeSpec, out changed);
+			default:
+				throw new NotImplementedException ($"Unknown type spec kind {type.Kind.ToString ()}");
+			}
+		}
+
+		TypeSpec Rename (ClosureTypeSpec type, out bool changed)
+		{
+			bool argsChanged, resultChanged;
+			var args = Rename (type.Arguments, out argsChanged);
+			var result = Rename (type.ReturnType, out resultChanged);
+			changed = argsChanged || resultChanged;
+			if (changed) {
+				return new ClosureTypeSpec (args, result);
+			}
+			return type;
+		}
+
+		TypeSpec Rename (NamedTypeSpec type, out bool changed)
+		{
+			bool nameChanged = false, genArgsChanged = false;
```

Comment:
Nit, I usually see this as two lines (though this is fine if you want to):

```suggestion
			bool nameChanged = false;
			bool genArgsChanged = false;
```